sandbox/apparmor/notify/message.go

```diff
@@ -0,0 +1,418 @@
+//	struct apparmor_notif {
+//	  struct apparmor_notif_common base;
+//	  __u16 ntype;        /* notify type */
+//	  __u8 signalled;
+//	  __u8 flags;
+//	  __u64 id;           /* unique id, not globally unique*/
+//	  __s32 error;        /* error if unchanged */
+//	} __attribute__((packed));
+type MsgNotification struct {
+	MsgHeader
+	// NotificationType describes the kind of notification message used.
+	// Currently the kernel only sends APPARMOR_NOTIF_OP messages.
+	NotificationType NotificationType
+	// XXX: Signaled seems to be unused.
+	Signalled uint8
+	// Set to 1 to NOT cache
+	Flags uint8
+	// ID is an opaque kernel identifier of the notification message. It must be
+	// repeated in the MsgNotificationResponse if one is sent back.
+	ID uint64
+	// XXX: This seems to be unused and clashes with identical field in
+	// MsgNotificationOp.
+	Error int32
+}
+
+// UnmarshalBinary unmarshals the message from binary form.
+func (msg *MsgNotification) UnmarshalBinary(data []byte) error {
+	const prefix = "cannot unmarshal apparmor notification message"
+
+	// Unpack the base structure.
+	if err := msg.MsgHeader.UnmarshalBinary(data); err != nil {
+		return err
+	}
+
+	// Unpack fixed-size elements.
+	buf := bytes.NewBuffer(data)
+	order := arch.Endian() // ioctl messages are native byte order, verify endianness if using for other messages
+	if err := binary.Read(buf, order, msg); err != nil {
+		return fmt.Errorf("%s: cannot unpack: %s", prefix, err)
+	}
+
+	return nil
+}
+
+// MarshalBinary marshals the message into binary form.
+func (msg *MsgNotification) MarshalBinary() ([]byte, error) {
+	msg.Version = 2
+	msg.Length = uint16(binary.Size(*msg))
+	buf := bytes.NewBuffer(make([]byte, 0, msg.Length))
+	order := arch.Endian() // ioctl messages are native byte order, verify endianness if using for other messages
+	if err := binary.Write(buf, order, msg); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// Validate returns an error if the mssage contains invalid data.
+func (msg *MsgNotification) Validate() error {
+	if !msg.NotificationType.IsValid() {
+		return fmt.Errorf("unsupported notification type: %d", msg.NotificationType)
+	}
+	return nil
+}
+
+// MsgNotificationResponse describes a response to a MsgNotification.
+//
+// This structure corresponds to the kernel type struct apparmor_notif
+// described below.
+//
+//	struct apparmor_notif_resp {
+//	  struct apparmor_notif base;
+//	  __s32 error;        /* error if unchanged */
+//	  __u32 allow;
+//	  __u32 deny;
+//	} __attribute__((packed));
+type MsgNotificationResponse struct {
+	MsgNotification
+	// XXX: The embedded MsgNotification also has an Error field, why?
+	Error int32
+	// Allow somehow encodes the allowed operation mask.
+	Allow uint32
+	// Deny somehow encodes the denied operation mask.
+	Deny uint32
+}
+
+// ResponseForRequest returns a response message for a given request.
+func ResponseForRequest(req *MsgNotification) MsgNotificationResponse {
```

Comment:
this seems not tested, also the implementation seems mostly full of XXXs? do we know what we require from it? is it premature?

Reply:
This is based on 3 structures, ideally we would move to a more optimized structure in v4. Comments add inline with ```#```

```
# common is required in all the messages sent over the ioctl interface
struct __attribute__((__packed__)) apparmor_notif_common {
	__u16 len;			/* actual len data */        #required
	__u16 version;		/* interface version */     # required
};

struct __attribute__((__packed__)) apparmor_notif {
	struct apparmor_notif_common base;                  # see above, required
	__u16 ntype;			/* notify type */              # set to APPARMOR_NOTIF_RESP
	__u8 signalled;                                                     # currently unused. In dev was used to signal some interrupt info
	__u8 reserved;                                                     # rename: to no_cache. set to 0 to cache response, 1 to not cache
	__u64 id;			/* unique id, not gloablly unique*/  # set to id of notif that is being replied to
	__s32 error;			/* error if unchanged */  # unused in responses.
};

/* userspace response to notification that expects a response */
struct __attribute__((__packed__)) apparmor_notif_resp {
	struct apparmor_notif base;
	__s32 error;			/* error if unchanged */  # unused in response.
	__u32 allow;                                                          # set to bit mask to be allowed
	__u32 deny;                                                          # set to bit mask to deny. Deny has priority
}
# deny | allow must cover at least the (allow | deny) in the original notification being replied to or the notification will result
# in a denial.  allow can be broader. Eg original request was for Read, reply with Read | Write. If used with caching a 
# subsequest request for Write will then be handle by the cache without sending a notification to userspace.
```
the two error fields were used in the v2 interface to allow setting/changing the return value from the kernel. But this allowed userspace to screw up between caching perms and the return codes resulting in odd failures. So this was taken away in v3.

Reply:
If my understanding is correct (and from your comments below), this means the error field of the `apparmor_notif` embedded in a response is the error which should be returned by the kernel to the application, and the error field in the `apparmor_notif_resp` is ignored?

Reply:
In v3 there is no error returned to the kernel in response. Both error fields (the one apparmor_notif embedded in response, and the error field in apparmor_notif_resp) are ignored.

For v3 these fields can be as marked as ignored within a response.